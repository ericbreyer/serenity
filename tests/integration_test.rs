#![cfg(test)]

use anyhow::{bail, Result};
use insta::assert_snapshot;
use test_case::{test_case, test_matrix};

#[test_case("trivial.ser" => matches Ok(1))]
#[test_case("prime_sieve.ser" => matches Ok(8181807856294299570))]
#[test_case("interfaces.ser" => matches Ok(12))]
#[test_case("babbage.ser" => matches Ok(25264))]
#[test_case("linkedlist.ser" => matches Ok(55))]
#[test_case("prime_conspiricy.ser" => matches Ok(7942686168))]
#[test_case("highly_composites.ser" => matches Ok(355168))]
#[test_case("generic.ser" => matches Ok(6))]
#[test_case("generic_func.ser" => matches Ok(540))]
#[test_case("generic_method.ser" => matches Ok(12))]
#[test_case("fib_stream.ser" => matches Ok(1346268))]
#[test_case("mat_fib.ser" => matches Ok(1346268))]
pub fn test_file_run(file: &str) -> Result<usize> {
    let file = format!("tests/test_files/{}", file);

    let mut out = Vec::new();
    let code = serenity::run_file(&file, &mut out)?;

    Ok(code as usize)
}

#[test_matrix(
    ["trivial.ser", "prime_sieve.ser", "interfaces.ser", "babbage.ser", "linkedlist.ser", "prime_conspiricy.ser", "highly_composites.ser", "generic.ser", "generic_func.ser", "generic_method.ser", "fib_stream.ser", "mat_fib.ser"],
    ["scan", "parse", "compile"]
)]
fn test_file_artifacts(file: &str, mode: &str) -> Result<()> {
    let file = format!("tests/test_files/{}", file);
    let result = match mode {
        "scan" => serenity::scan(&file)?
            .iter()
            .fold(String::new(), |acc, x| acc + &format!("{}\n", x)),
        "parse" => serenity::parse(&file)?,
        "compile" => serenity::compile(&file)?,
        _ => bail!("Invalid mode"),
    };

    let mut settings = insta::Settings::clone_current();
    settings.set_snapshot_suffix(format!("{}_{}", file, mode));
    let _g = settings.bind_to_scope();

    assert_snapshot!(result);

    Ok(())
}
