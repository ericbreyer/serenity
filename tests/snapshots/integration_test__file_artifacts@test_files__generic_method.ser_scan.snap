---
source: tests/integration_test.rs
expression: result
---
<Type         "type"        [1]>
<Less         "<"           [1]>
<Identifier   "T"           [1]>
<Greater      ">"           [1]>
<Identifier   "test"        [1]>
<Struct       "struct"      [1]>
<LeftBrace    "{"           [1]>
<Identifier   "a"           [2]>
<Colon        ":"           [2]>
<Identifier   "T"           [2]>
<Comma        ","           [2]>
<Identifier   "b"           [3]>
<Colon        ":"           [3]>
<Identifier   "T"           [3]>
<Comma        ","           [3]>
<RightBrace   "}"           [4]>
<Semicolon    ";"           [4]>
<Fun          "fn"          [6]>
<Less         "<"           [6]>
<Identifier   "T"           [6]>
<Greater      ">"           [6]>
<LeftParen    "("           [6]>
<Identifier   "t"           [6]>
<Colon        ":"           [6]>
<Star         "*"           [6]>
<Struct       "struct"      [6]>
<Identifier   "test"        [6]>
<Less         "<"           [6]>
<Identifier   "T"           [6]>
<Greater      ">"           [6]>
<RightParen   ")"           [6]>
<Identifier   "add"         [6]>
<LeftParen    "("           [6]>
<Identifier   "c"           [6]>
<Colon        ":"           [6]>
<Identifier   "T"           [6]>
<RightParen   ")"           [6]>
<RightArrow   "->"          [6]>
<Identifier   "T"           [6]>
<LeftBrace    "{"           [6]>
<Return       "return"      [7]>
<Identifier   "t"           [7]>
<RightArrow   "->"          [7]>
<Identifier   "a"           [7]>
<Plus         "+"           [7]>
<Identifier   "t"           [7]>
<RightArrow   "->"          [7]>
<Identifier   "b"           [7]>
<Plus         "+"           [7]>
<Identifier   "c"           [7]>
<Semicolon    ";"           [7]>
<RightBrace   "}"           [8]>
<Fun          "fn"          [10]>
<Identifier   "main"        [10]>
<LeftParen    "("           [10]>
<RightParen   ")"           [10]>
<RightArrow   "->"          [10]>
<SimpleType   "int"         [10]>
<LeftBrace    "{"           [10]>
<Var          "var"         [11]>
<Identifier   "d"           [11]>
<Colon        ":"           [11]>
<Struct       "struct"      [11]>
<Identifier   "test"        [11]>
<Less         "<"           [11]>
<SimpleType   "int"         [11]>
<Greater      ">"           [11]>
<Semicolon    ";"           [11]>
<Identifier   "d"           [12]>
<Dot          "."           [12]>
<Identifier   "a"           [12]>
<Equal        "="           [12]>
<Number       "1"           [12]>
<Semicolon    ";"           [12]>
<Identifier   "d"           [13]>
<Dot          "."           [13]>
<Identifier   "b"           [13]>
<Equal        "="           [13]>
<Number       "2"           [13]>
<Semicolon    ";"           [13]>
<Var          "var"         [15]>
<Identifier   "f"           [15]>
<Colon        ":"           [15]>
<Struct       "struct"      [15]>
<Identifier   "test"        [15]>
<Less         "<"           [15]>
<SimpleType   "float"       [15]>
<Greater      ">"           [15]>
<Semicolon    ";"           [15]>
<Identifier   "f"           [16]>
<Dot          "."           [16]>
<Identifier   "a"           [16]>
<Equal        "="           [16]>
<Number       "1.0"         [16]>
<Semicolon    ";"           [16]>
<Identifier   "f"           [17]>
<Dot          "."           [17]>
<Identifier   "b"           [17]>
<Equal        "="           [17]>
<Number       "2.0"         [17]>
<Semicolon    ";"           [17]>
<Return       "return"      [19]>
<Identifier   "d"           [19]>
<Dot          "."           [19]>
<Identifier   "add"         [19]>
<LeftParen    "("           [19]>
<Number       "3"           [19]>
<RightParen   ")"           [19]>
<Plus         "+"           [19]>
<SimpleType   "int"         [19]>
<LeftParen    "("           [19]>
<Identifier   "f"           [19]>
<Dot          "."           [19]>
<Identifier   "add"         [19]>
<LeftParen    "("           [19]>
<Number       "3.0"         [19]>
<RightParen   ")"           [19]>
<RightParen   ")"           [19]>
<Semicolon    ";"           [19]>
<RightBrace   "}"           [20]>
<Eof          ""            [20]>
